feat: add `utils/any-in-by`

itsspriyansh opened the pull request:
Resolves #823 

## Description

> What is the purpose of this pull request?

This pull request:

-   a adds utils/any-in-by package
-   b tests whether any "own" and inherited property of a provided object satisfies a predicate function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [ ] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

itsspriyansh committed:
feat: added @stdlib/utils/any-in-by

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
````diff
@@ -0,0 +1,252 @@
+var anyInBy = require( '@stdlib/utils/any-in-by' );
+```
+
+#### anyInBy( collection, predicate\[, thisArg ] )
````
```suggestion
#### anyInBy( object, predicate\[, thisArg ] )
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
```diff
@@ -0,0 +1,252 @@
+-   <span class="package-name">[`@stdlib/utils/every-by`][@stdlib/utils/every-by]</span><span class="delimiter">: </span><span class="description">test whether all elements in a collection pass a test implemented by a predicate function.</span>
+-   <span class="package-name">[`@stdlib/utils/for-each`][@stdlib/utils/for-each]</span><span class="delimiter">: </span><span class="description">invoke a function for each element in a collection.</span>
+-   <span class="package-name">[`@stdlib/utils/none-by`][@stdlib/utils/none-by]</span><span class="delimiter">: </span><span class="description">test whether all elements in a collection fail a test implemented by a predicate function.</span>
+-   <span class="package-name">[`@stdlib/utils/some-by`][@stdlib/utils/some-by]</span><span class="delimiter">: </span><span class="description">test whether a collection contains at least `n` elements which pass a test implemented by a predicate function.</span>
```
We have tooling to auto-populate this section, so no reason to fill in manually.

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
```diff
@@ -0,0 +1,252 @@
+
+[@stdlib/utils/none-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/none-by
+
+[@stdlib/utils/some-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/some-by
```
We have tooling to auto-populate these links, so no reason to fill in manually.

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
```diff
@@ -0,0 +1,252 @@
+
+-   `value`: collection element
+-   `index`: collection index
+-   `collection`: input collection
```
```suggestion
-   `value`: property value
-   `key`: property key
-   `object`: input object
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
```diff
@@ -0,0 +1,252 @@
+
+-   The `obj` object may contain own properties as well as inherited properties from its prototype chain.
+
+-   If provided an empty collection, the function returns `false`.
```
```suggestion
-   If provided an empty object, the function returns `false`.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/README.md:
````diff
@@ -0,0 +1,252 @@
+        */
+        ```
+
+    -   The function provides limited support for dynamic collections (i.e., collections whose `length` changes during execution).
````
Let's remove these notes as specifically about arrays.

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/repl.txt:
```diff
@@ -0,0 +1,41 @@
+{{alias}}( collection, predicate[, thisArg ] )
```
REPL documentation here needs to be updates as it currently is written for `collection` / `array` inputs, but the function is for `object`s.

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/test/test.js:
```diff
@@ -0,0 +1,132 @@
+	}
+});
+
+tape( 'the function throws an error if not provided a object', function test( t ) {
+	var values;
+	var i;
+
+	values = [
+		'5',
+		5,
+		NaN,
+		true,
+		false,
+		null,
+		void 0,
+		[],
+		function noop() {},
+		/.*/,
+		new Date()
+	];
+	for (i =0; i < values.length; i++) {
+		t.throws( badValue( values ), TypeError, 'throws a type error when provided '+values[i] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			anyInBy( value, noop );
+		};
+	}
+});
```
Can remove as duplicate of test in L36-63.

```suggestion
```

However, we need a test for ensuring that function throws an error if provided a predicate argument that is not a function.

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/package.json:
```diff
@@ -0,0 +1,69 @@
+      "util",
+      "test",
+      "predicate",
+      "any",
+      "array.some",
+      "iterate",
+      "collection",
+      "array-like",
```
```suggestion
      "object",
      "obj",
      "iterate",
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/lib/main.js:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	for ( key in obj ) {
+		if ( hasOwnProp( obj, key ) && predicate.call( thisArg, obj[key], key, obj ) ) {
```
We should remove the `hasOwnProp( obj, key )` property check as the function is for "own _and inherited_" properties.

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+* // returns true
+*/
+declare function anyInBy<T = unknown, U = unknown>( obj: { [key: string]: T }, predicate: Predicate<T, U>, thisArg?: ThisParameterType<Predicate<T, U>> ): boolean;
+
+// EXPORTS //
```
Two empty lines before section header comments:

```suggestion


// EXPORTS //
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+import { Collection } from '@stdlib/types/array';
+
+/**
+* Checks whether an element in a collection passes a test.
```
```suggestion
* Checks whether an object property passes a test.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @returns boolean indicating whether an element in a collection passes a test
```
```suggestion
* @returns boolean indicating whether an object property passes a test
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+type Nullary<U> = ( this: U ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
```
```suggestion
* Checks whether an object property passes a test.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @returns boolean indicating whether an element in a collection passes a test
```
```suggestion
* @returns boolean indicating whether an object property passes a test
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+type Unary<T, U> = ( this: U, value: T ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
```
```suggestion
* Checks whether an object property passes a test.
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
```
```suggestion
* @param value - property value
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
```
```suggestion
* @param value - property value
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @param index - collection index
```
```suggestion
* @param key - property key
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+*
+* @param value - collection value
+* @param index - collection index
+* @returns boolean indicating whether an element in a collection passes a test
```
```suggestion
* @returns boolean indicating whether an object property passes a test
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+type Binary<T, U> = ( this: U, value: T, index: number ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @param index - collection index
+* @param collection - input collection
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Ternary<T, U> = ( this: U, value: T, index: number, collection: Collection<T> ) => boolean;
```
```suggestion
* Checks whether an object property passes a test.
*
* @param value - property value
* @param key - property key
* @param obj - input object
* @returns boolean indicating whether an object property passes a test
*/
type Ternary<T, U> = ( this: U, value: T, key: string, obj:{ [key: string]: T } ) => boolean;
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:
```diff
@@ -0,0 +1,103 @@
+*/
+type Ternary<T, U> = ( this: U, value: T, index: number, collection: Collection<T> ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @param index - collection index
+* @param collection - input collection
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Predicate<T, U> = Nullary<U> | Unary<T, U> | Binary<T, U> | Ternary<T, U>;
```
```suggestion
* Checks whether an object property passes a test.
*
* @param value - property value
* @param key - property key
* @param obj - input object
* @returns boolean indicating whether an object property passes a test
*/
type Predicate<T, U> = Nullary<U> | Unary<T, U> | Binary<T, U> | Ternary<T, U>;
```

Planeshifter reviewed (changes requested):
Thanks for this PR! Left some comments that will need to be addressed before this can be merged.

itsspriyansh commented:
@Planeshifter I have resolved all the problems.

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/test/test.js:
```diff
@@ -0,0 +1,103 @@
+	t.end();
+});
+
+tape( 'the function throws an error if not provided a object', function test( t ) {
```
```suggestion
tape( 'the function throws an error if not provided an object', function test( t ) {
```

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/test/test.js:
```diff
@@ -0,0 +1,103 @@
+		};
+	}
+});
+
```
Needs a test for when `predicate` is not a function.

Planeshifter commented on lib/node_modules/@stdlib/utils/any-in-by/test/test.js:
```diff
@@ -0,0 +1,103 @@
+
+	t.strictEqual( bool, true, 'returns true' );
+	t.end();
+});
```
We should have a test for when the function returns `false`. Our goal is to achieve 100% code coverage, so we should make sure that tests cover all code paths.

Planeshifter reviewed (commented):
It's coming along well, but I noticed we are still missing some tests. Would you be able to please add these?

itsspriyansh committed:
added some more tests

Signed-off-by: Priyansh <88396544+itsspriyansh@users.noreply.github.com>

itsspriyansh committed:
added some tests

Signed-off-by: Priyansh <88396544+itsspriyansh@users.noreply.github.com>

itsspriyansh commented:
@Planeshifter I have fixed the typo and I have added 3 tests where:
1. predicate is not a function
2. predicate is not provided
3. function returns false

Planeshifter reviewed (approved):
LGTM! Thanks for this contribution!
